guides/contributing.rst

```diff
@@ -592,83 +693,83 @@ if communication works (if not, it should call ``mark_failed()``).
 
 Again, look at examples of other components to learn more.
 
-The next thing that will be called with your component is ``loop()`` (or ``update()`` for a
-:apiclass:`PollingComponent`). In these methods you should retrieve the latest data from the
-component and publish them with the provided methods. One thing to note in these methods
-is that anything in ``loop()`` or ``setup()`` **should not block**. Specifically methods like
-``delay(10)`` should be avoided and delays above ~10ms are not permitted. The reason for this
-is that ESPHome uses a central single-threaded loop for all components - if your component
-blocks the whole loop will be slowed down.
+The next method that will be called with your component is ``loop()`` (or ``update()`` for a
+:apiclass:`PollingComponent`). These methods should retrieve the latest data from your component and publish them with
+the provided methods.
 
-Finally, your component should have a ``dump_config`` method that prints the user configuration.
+.. note::
+
+    **Code in** ``loop()``, ``update()`` **and** ``setup()`` **must not block**. Specifically, methods like ``delay()``
+    should be avoided and **delays longer than 10 ms are not permitted**. The reason for this is that ESPHome uses a
+    single-threaded loop for all components -- if your component blocks, it will delay the whole loop, negatively
+    impacting other components. This can result in a variety of problems such as network connections being lost.
```

Comment:
Maybe here give reference to `this->set_timeout` or using a state-machine to handle "waiting" in loop for the state to become what is needed for the next stage?

Reply:
See how this looks: [`33d4838` (#4188)](https://github.com/esphome/esphome-docs/pull/4188/commits/33d48387bcba798c6ea072ee5db0a5ebed32f61d)